sycl/doc/extensions/proposed/sycl_ext_oneapi_global_defaults.asciidoc

```diff
@@ -0,0 +1,977 @@
+----
+
+a|
+[frame=all,grid=none]
+!====
+a!
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+template <typename T>
+void copy(const T* src, T* dest, size_t count);
+
+}
+----
+!====
+_Effects_: The same result as computed by:
+
+[source, c++]
+----
+copy(get_current_device().ext_oneapi_get_default_queue(), src, dest, count)
+----
+
+a|
+[frame=all,grid=none]
+!====
+a!
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+void memset(void* ptr, int value, size_t numBytes);
+
+}
+----
+!====
+_Effects_: The same result as computed by:
+
+[source, c++]
+----
+memset(get_current_device().ext_oneapi_get_default_queue(), ptr, value, numBytes)
+----
+
+a|
+[frame=all,grid=none]
+!====
+a!
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+template <typename T>
+void fill(T* ptr, const T& pattern, size_t count);
+
+}
+----
+!====
+_Effects_: The same result as computed by:
+
+[source, c++]
+----
+fill(get_current_device().ext_oneapi_get_default_queue(), ptr, pattern, count)
+----
+
+a|
+[frame=all,grid=none]
+!====
+a!
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+void prefetch(void* ptr, size_t numBytes);
+
+}
+----
+!====
+_Effects_: The same result as computed by:
+
+[source, c++]
+----
+prefetch(get_current_device().ext_oneapi_get_default_queue(), ptr, numBytes)
+----
+
+a|
+[frame=all,grid=none]
+!====
+a!
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+void mem_advise(void* ptr, size_t numBytes, int advice);
+
+}
+----
+!====
+_Effects_: The same result as computed by:
+
+[source, c++]
+----
+mem_advise(get_current_device().ext_oneapi_get_default_queue(), ptr, numBytes, advice)
+----
+
+|====
+
+
+==== Command barriers
+
+The functions in this section are only available if the
+link:../supported/sycl_ext_oneapi_enqueue_barrier.asciidoc[
+  sycl_ext_oneapi_enqueue_barrier] extension is supported.
+
+|====
+a|
+[frame=all,grid=none]
+!====
+a!
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+void barrier();
+
+}
+----
+!====
+_Effects_: The same result as computed by:
+
+[source, c++]
+----
+barrier(get_current_device().ext_oneapi_get_default_queue())
+----
+
+a|
+[frame=all,grid=none]
+!====
+a!
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+void partial_barrier(const std::vector<sycl::event>& events);
+
+}
+----
+!====
+_Effects_: The same result as computed by:
+
+[source, c++]
+----
+partial_barrier(get_current_device().ext_oneapi_get_default_queue(), events)
+----
+
+[_Note:_ If `events` is empty and a partial barrier has no other dependencies
+(e.g., specified by `handler::depends_on`), it is not required to wait for any
+commands unless the `queue` is in-order. Implementations may be able to
+optimize such partial barriers.
+_{endnote}_]
+|====
+
+== Issues
+. [UNRESOLVED] Should the currrent device be global or should we also support a per-thread
+   device?
```

Comment:
FWIW, the CUDA versions are always per-thread: https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html?highlight=cudaSetDevice#device-selection

I can see this behavior being useful for something like an OpenMP+SYCL application -- you could call `set_default_device()` on each OpenMP thread to round-robin assign your devices to threads, then not worry about it again.

Reply:
Some thoughts here that are relevant if we decided that each thread has its own default device:

* We should clarify the behavior when a new thread is created.  Does the new thread inherit the parent thread's default device?  If not, is the initial default device for the new thread guaranteed to be the device selected by the default selector?  Or, is the initial default device for a new thread unspecified?

* We have discussed in the past the possibility of expanding the set of APIs that can be called from a host task.  It is possible that we will allow a host task to call some of the APIs in this extension.  If that happens, what guarantees does the application have about the default device in a host task?